README.md
# yikesbot

**yikesbot** is a Destiny.gg chat bot that has an integrated "yikes" counter that increases every time it detects a variation of the word "yikes" in a chat message. It is heavily inspired by the [RandomFerret](https://github.com/voloshink/FerretBot) bot made by **Polecat** in chat. Most of his existing code was used in creating this bot.

## How does it work

yikesbot listenes for a defined variation of the word "yikes" in a chat message and increases its internal counter by 10. The internal counter decreases by 1 every 2 seconds until it reaches 0.

Expected variations of "yikes" the bot currently processes:


- Yikes
- Y I K E S
- Yikerz
- Yikers

Capitalization is not important. Any of the above words can be spelled as you wish.

### New since Version 1.5

- omegayikes

Typing omegayikes increases the current amount by 50 but has a cooldown of 60 seconds before it can be used again.

### New since Version 1.7.1

- !graphyikes links to a graphical overview that displays the yikes amount over time.

## Commands

## **UPDATE**: Yikesbot got banned and is now only operating in stealth mode... No more !yikes

Currently everyone in chat can get the current yikes count by typing the following commands:

- !yikes
- !YIKES
- ! yikes
- ! YIKES
- !yikers
- !YIKERS

Cooldown of the command is currently set to 10 seconds

*Note that the command just needs to start with one of the above words. You can append any word behind a command: i.e. !yikes**BOI***

## Admin Commands

Currently theneedyguy and LeEpicMemeXd are admins of the bot and can thus execute the following commands:

| Command         | Effect                                                                  |
| ----------      |:----------------------------------------------------------------------- |
| **!ver**        | Displays current version of yikesbot                                    |
| **!reset**      | Resets the yikes counter                                                |
| **!sleep**      | Toggles Sleep Mode (Bot stops responding to !yikes)                     |
| **!ping**       | Sends back a pong (Check if bot is alive)                               |
| **!ipban**      | Increases the counter by 100 (Triggered only when Destiny bans someone) |
| **!topyikes**   | Shows the highest yikes level ever achieved since starting the bot      |

## Metrics

Since version 1.7 yikesbot exposes Prometheus time series data on port 9998. This is only relevant for people that know what Prometheus is and what it is used for.
Metric label:

- dgg_yikes_level

## Changelog

[Changelogs are here](https://github.com/theneedyguy/yikesbot/blob/master/CHANGELOG.md)

## Feature plans

- Some kind of "buy" function for yikes. Will require a database or just a local json that gets updated. We'll see what I can come up with.
The buy function would come with a cooldown. The first person to buy yikes would get them. A leaderboard would be a nice meme.


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/theneedyguy/yikesbot/compare/v1.7.3...HEAD)

## [1.7.3](https://github.com/theneedyguy/yikesbot/compare/v1.7.2...v1.7.3) - 2018-03-10

### Changes

- Stealth mode due to ban... Only counts yikes never reacts to the command.

## [1.7.2](https://github.com/theneedyguy/yikesbot/compare/v1.7.1...v1.7.2) - 2018-02-15

### Changes

- Fixes graph command

## [1.7.1](https://github.com/theneedyguy/yikesbot/compare/v1.7.0...v1.7.1) - 2018-02-15

### Added

- Command to link to graphical overview
- URL to graphical overview is configured via the config file

## [1.7](https://github.com/theneedyguy/yikesbot/compare/v1.6.1...v1.7.0) - 2018-02-15

### Added

- Added Prometheus metrics endpoint for time series data. Will be used for graphing yikes level over time.

## [1.6.1](https://github.com/theneedyguy/yikesbot/compare/v1.6.0...v1.6.1) - 2018-02-15

### Added

- "Yikerz" is now recognized by the bot.

### Changed

- Got rid of some excess code.
- Yikes are recognized regardless of the capitalization

## [1.6](https://github.com/theneedyguy/yikesbot/compare/v1.5.0...v1.6.0) - 2018-02-15

### Changed

- When outputting the yikes level the emote will vary depending on the amount of the current yikes.

## 1.5.1-2 - 2018-02-14

### Changed

- Fixed problems with ipban not getting detected.

## 1.5.1- 2018-02-14

### Added

- Different messages when Destiny bans someone. Prevents duplicate message error.

## [1.5.0](https://github.com/theneedyguy/yikesbot/tree/v1.5.0) - 2018-02-11

### Added

- Bot now counts "omegayikes" or "OMEGAYIKES" as 50 points. Cooldown is 1 minute.
- Added rosewood18 to list of admins

### Changed

- Changed a function to raise the amount of yikes

### Removed

- Removed some comments

## 1.4.0 - 2018-02-09

### Added

- New admin command: **!topyikes** (Shows the highest yikes level ever achieved since starting the bot)

## 1.3.0 - 2018-02-09

### Added

- Added all features
- Admin Commands
- Default Commands

commands.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/voloshink/dggchat"
)

func handleCommand(m dggchat.Message, s *dggchat.Session) {

	if isYikesCommand(m.Message) {
		//handleYikesCommand(m, s)
		return
	}

	if strings.HasPrefix(m.Message, "!reset") {
		if isAdmin(m.Sender.Nick) {
			resetYikes()
			_ = s.SendMessage("YIKER-COUNTER RESET FeelsBadMan")
		}
	}
	/*
		if strings.HasPrefix(m.Message, "!graphyikes") {
			timeElapsed := time.Since(lastSent)
			if timeElapsed < messageInterval {
				return
			}
			switch graphMessage {
			case 0:
				_ = s.SendMessage(fmt.Sprintf("Graphical Yikes: %s DANKMEMES", graphURL))
				graphMessage = graphMessage + 1
			case 1:
				_ = s.SendMessage(fmt.Sprintf("Yikes Graph: %s DANKMEMES", graphURL))
				graphMessage = graphMessage - 1
			default:
				_ = s.SendMessage(fmt.Sprintf("Graphical Yikes: %s DANKMEMES", graphURL))
				graphMessage = graphMessage + 1
			}
			lastSent = time.Now()
		}
	*/
	if strings.HasPrefix(m.Message, "!topyikes") {
		if isAdmin(m.Sender.Nick) {
			_ = s.SendMessage(fmt.Sprintf("Highest recorded yikes level (since starting the bot): %s", strconv.Itoa(yikesTop)))
		}
	}

	if strings.HasPrefix(m.Message, "!sleep") {

		//if strings.EqualFold("theneedyguy", m.Sender.Nick) {
		if isAdmin(m.Sender.Nick) {
			isSleeping := toggleSleep()
			if isSleeping == true {
				_ = s.SendMessage("I am going to sleep...")
			} else {
				_ = s.SendMessage("Ready to count yikes again.")
			}

		}
	}

	if strings.HasPrefix(m.Message, "!ipban") {
		if strings.EqualFold("Destiny", m.Sender.Nick) {
			//fmt.Println("IPBAN DETECTED!")
			switch ipbanMessage {
			case 0:
				_ = s.SendMessage("Yikes increased by 100 due to ban.")
				raiseYikesLevel(100)
				//fmt.Println("IP1")
				ipbanMessage = ipbanMessage + 1
			case 1:
				_ = s.SendMessage("Ban! 100 added to yikes-o-meter.")
				raiseYikesLevel(100)
				//fmt.Println("IP2")
				ipbanMessage = ipbanMessage - 1
			default:
				_ = s.SendMessage("Ban! 100 added to yikes-o-meter.")
				raiseYikesLevel(100)
				ipbanMessage = 0
				//fmt.Println("IPDEF")
			}

		}
	}

	if strings.HasPrefix(m.Message, "!ver") {
		if isAdmin(m.Sender.Nick) {
			_ = s.SendMessage(yikesVersion)
		}
	}
	if strings.HasPrefix(m.Message, "!ping") {
		if isAdmin(m.Sender.Nick) {
			_ = s.SendMessage("YIKESPONG")
		}
	}
}

func isYikesCommand(s string) bool {
	for _, c := range yikesCommands {
		if strings.HasPrefix(s, c) {
			return true
		}
	}
	return false
}

func setTopYikes(yikesLevelInput int) {
	if yikesLevelInput > yikesTop {
		yikesTop = yikesLevelInput
	}
}

func resetYikes() {
	yikesLevel = 0
}

func toggleSleep() bool {
	if yikesSleep == false {
		yikesSleep = true
		return true
	}
	yikesSleep = false
	return false
}

func getYikesLevel() (string, error) {
	switch yikesMessage {
	case 0:
		message := []string{"AMOUNT OF YIKES:", strconv.Itoa(yikesLevel)}
		yikesMessage = yikesMessage + 1
		return strings.Join(message, " "), nil
	case 1:
		message := []string{"LEVELS OF YIKES:", strconv.Itoa(yikesLevel)}
		yikesMessage = yikesMessage + 1
		return strings.Join(message, " "), nil
	case 2:
		message := []string{"TOTAL YIKES:", strconv.Itoa(yikesLevel)}
		yikesMessage = yikesMessage - 2
		return strings.Join(message, " "), nil
	default:
		message := []string{"CHAT YIKES:", strconv.Itoa(yikesLevel)}
		yikesMessage = 0
		return strings.Join(message, " "), nil
	}
}

func raiseYikesLevel(amount int) {
	yikesLevel = yikesLevel + amount
	setTopYikes(yikesLevel)
}

func omegaYikes() {
	omegaYikesElapsed := time.Since(lastOmegaYikes)
	if omegaYikesElapsed < omegaYikesinterval {
		return
	}
	raiseYikesLevel(50)
	lastOmegaYikes = time.Now()
}

func decreaseYikesOverTime() {
	for {
		//goroutine that runs indefinitely.
		if yikesLevel != 0 {
			yikesLevel = yikesLevel - 1
		}
		yikesMetric.Set(float64(yikesLevel))
		time.Sleep(time.Second * 2)
	}
}

func isAdmin(s string) bool {
	return isInList(s, admins)
}

func isInList(s string, list []string) bool {
	for _, v := range list {
		if strings.EqualFold(s, v) {
			return true
		}
	}
	return false
}

func handleYikesCommand(m dggchat.Message, s *dggchat.Session) {
	timeElapsed := time.Since(lastSent)
	if timeElapsed < messageInterval {
		return
	}

	ylevel, err := getYikesLevel()
	if err != nil {
		return
	}
	/*
		if ylevel == lastMessage {
			return
		}
	*/
	if yikesSleep == false {
		switch {
		case yikesLevel < 100:
			err = s.SendMessage(fmt.Sprintf("%s SLEEPSTINY", ylevel))
			if err != nil {
				return
			}
		case yikesLevel > 100 && yikesLevel < 500:
			err = s.SendMessage(fmt.Sprintf("%s Memegasm", ylevel))
			if err != nil {
				return
			}
		case yikesLevel > 500 && yikesLevel < 1000:
			err = s.SendMessage(fmt.Sprintf("%s WhoahDude", ylevel))
			if err != nil {
				return
			}
		case yikesLevel > 1000:
			err = s.SendMessage(fmt.Sprintf("%s GODSTINY", ylevel))
			if err != nil {
				return
			}
		}
	} else {
		err = s.SendMessage(fmt.Sprintf("%s I am sleeping until awakened by my master.", "SLEEPSTINY"))
		if err != nil {
			return
		}
	}
	lastSent = time.Now()
	lastMessage = ylevel
}


main.go
package main

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/voloshink/dggchat"
)

const (
	defaultConfigFile = "config.json"
	pingInterval      = time.Minute
)

var (
	yikesCommands      = []string{"!yikes", "!YIKES", "! yikes", "! YIKES", "!yikers", "!YIKERS"}
	lastMessage        = ""
	yikesVersion       = "1.7.3"
	yikesLevel         = 0
	yikesMessage       = 0
	ipbanMessage       = 0
	graphMessage       = 0
	yikesSleep         = false
	yikesTop           = 0
	lastOmegaYikes     = time.Now()
	omegaYikesinterval = time.Second * 60
	lastSent           = time.Now()
	lastPM             = time.Now()
	startTime          = time.Now()
	lastPing           = timeToUnix(time.Now())
	lastPong           = timeToUnix(time.Now())
	messageInterval    = time.Second * 10
	pmInterval         = time.Second * 30
	configFile         string
	admins             []string
	graphURL           string
	key                string
	yikesMetric        = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "dgg_yikes_level",
		Help: "Current Yikes Level",
	})
)

type (
	config struct {
		Key    string   `json:"login_key"`
		Admins []string `json:"admins"`
		Graph  string   `json:"graph"`
	}

	apiResp struct {
		URL string `json:"url"`
	}
)

func main() {
	var file string
	if len(os.Args) < 2 {
		file = defaultConfigFile
	} else {
		file = os.Args[1]
	}

	f, err := ioutil.ReadFile(file)
	if err != nil {
		log.Fatalln(err)
	}

	var c config
	err = json.Unmarshal(f, &c)
	if err != nil {
		log.Fatalln(err)
	}

	configFile = file

	if c.Key == "" {
		log.Fatalln("No login key provided")
	}

	key = c.Key

	if c.Admins == nil {
		c.Admins = make([]string, 0)
	}

	admins = c.Admins

	if c.Graph == "" {
		log.Fatalln("No Graph URL provided")
	}

	graphURL = c.Graph

	go startBot(c.Key)
	go runMetrics()
	go decreaseYikesOverTime()

	sc := make(chan os.Signal, 1)
	signal.Notify(sc, syscall.SIGINT)
	<-sc
}

func startBot(key string) {
	dgg, err := dggchat.New(key)
	if err != nil {
		log.Fatalln(err)
	}

	err = dgg.Open()
	if err != nil {
		log.Fatalln(err)
	}

	defer dgg.Close()

	messages := make(chan dggchat.Message)
	errors := make(chan string)
	pings := make(chan dggchat.Ping)

	dgg.AddMessageHandler(func(m dggchat.Message, s *dggchat.Session) {
		messages <- m
	})

	dgg.AddErrorHandler(func(e string, s *dggchat.Session) {
		errors <- e
	})

	dgg.AddPingHandler(func(p dggchat.Ping, s *dggchat.Session) {
		pings <- p
	})

	go checkConnection(dgg)

	for {
		select {
		case m := <-messages:
			if strings.Contains(strings.ToLower(m.Message), "omegayikes") {
				omegaYikes()
			}
			if strings.Contains(strings.ToLower(m.Message), "yikes") ||
				strings.Contains(strings.ToLower(m.Message), "y i k e s") ||
				strings.Contains(strings.ToLower(m.Message), "yikers") ||
				strings.Contains(strings.ToLower(m.Message), "yikerz") {
				raiseYikesLevel(10)
			}
			if strings.HasPrefix(m.Message, "!") {
				handleCommand(m, dgg)
			}
		case e := <-errors:
			log.Printf("Error %s\n", e)
		case p := <-pings:
			lastPong = p.Timestamp
		}
	}

}

func checkConnection(s *dggchat.Session) {
	ticker := time.NewTicker(pingInterval)
	for {
		<-ticker.C
		if lastPing != lastPong {
			log.Println("Ping mismatch, attempting to reconnect")
			err := s.Close()
			if err != nil {
				log.Fatalln(err)
			}

			err = s.Open()
			if err != nil {
				log.Fatalln(err)
			}

			continue
		}
		s.SendPing()
		lastPing = timeToUnix(time.Now())
	}
}

func timeToUnix(t time.Time) int64 {
	return t.Unix() * 1000
}
